src/main/angular/app/services/graph-processor.service.ts

```diff
@@ -0,0 +1,275 @@
+      {type: 'string', role: 'style'}
+    );
+  }
+
+  /** From a given SimpleChange, extract the projects that were added or deleted. */
+  private getAdditionsDeletions(
+    change: SimpleChange
+  ): {added: Project[]; removed: Project[]} {
+    const out: {added: Project[]; removed: Project[]} = {
+      added: [],
+      removed: [],
+    };
+    if (change.previousValue === undefined) {
+      out.added = change.currentValue;
+      return out;
+    }
+
+    // Look for additions
+    change.currentValue
+      .filter((c: Project) => !change.previousValue.includes(c))
+      .forEach((addition: Project) => out.added.push(addition));
+    // Look for removals
+    change.previousValue
+      .filter((c: Project) => !change.currentValue.includes(c))
+      .forEach((deletion: Project) => out.removed.push(deletion));
+    return out;
+  }
+
+  /** Returns the tooltip associated with the given IAM Bindings time */
+  private getTooltip(
+    numberBindings: number,
+    matchingRecommendations: Recommendation[]
+  ): string {
+    // The list of recommendations on the same day
+    if (matchingRecommendations.length === 0) {
+      return `IAM Bindings: ${numberBindings}`;
+    }
+
+    let tooltip = '';
+    matchingRecommendations.forEach((recommendation, index) => {
+      tooltip += recommendation.description;
+      if (index < matchingRecommendations.length - 1) {
+        tooltip += '\n';
+      }
+    });
+    return tooltip;
+  }
+
+  /** Returns the point styling associated with the given recommendation. */
+  private getPoint(
+    matchingRecommendations: Recommendation[],
+    color: string
+  ): string | undefined {
+    if (matchingRecommendations.length === 0) {
+      return undefined;
+    }
+    return `point { size: 10; shape-type: circle; fill-color: ${color}; visible: true; }`;
+  }
+
+  /** Returns the recommendations which occured on the same day as the given time, which is in milliseconds since epoch. */
+  private getRecommendationsOnSameDay(
+    time: number,
+    dateToRecommendation: {[timeInMillis: number]: Recommendation}
+  ): Recommendation[] {
+    const recommendations: Recommendation[] = [];
+    for (const [key, value] of Object.entries(dateToRecommendation)) {
+      if (this.dateUtilities.fallOnSameDay(time, +key)) {
+        recommendations.push(value);
+      }
+    }
+    return recommendations;
+  }
+
+  /** Checks if the rows already includes the given day. */
+  private includesDay(rows: Row[], day: Date): boolean {
+    return (
+      rows.findIndex(row => {
+        if (row[0] instanceof Date) {
+          return row[0].getTime() === day.getTime();
+        }
+        return false;
+      }) !== -1
+    );
+  }
+
+  /** Returns the row representing the given day. */
+  private getRow(rows: Row[], day: Date): Row | undefined {
+    return rows.find(row => {
+      if (row[0] instanceof Date) {
+        return row[0].getTime() === day.getTime();
+      }
+      return false;
```

Comment:
is there a meaningful difference between returning null vs false?

Reply:
The return false is still inside the `find` call. It indicates that that particular row is not what we're looking for.